flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java

```diff
@@ -0,0 +1,407 @@
+    /**
+     * Completed stats. Important: Job vertex IDs need to be scoped by job ID, because they are
+     * potentially constant across runs messing up the cached data.
+     */
+    @GuardedBy("lock")
+    private final Cache<AccessExecutionJobVertex, T> operatorStatsCache;
+
+    /**
+     * Pending in progress stats. Important: Job vertex IDs need to be scoped by job ID, because
+     * they are potentially constant across runs messing up the cached data.
+     */
+    @GuardedBy("lock")
+    private final Set<AccessExecutionJobVertex> pendingStats = new HashSet<>();
+
+    private final int numSamples;
+
+    private final int statsRefreshInterval;
+
+    private final Time delayBetweenSamples;
+
+    private final int maxThreadInfoDepth;
+
+    // Used for testing purposes
+    private final CompletableFuture<Void> resultAvailableFuture = new CompletableFuture<>();
+
+    /** Flag indicating whether the stats tracker has been shut down. */
+    private boolean shutDown;
+
+    private ThreadInfoOperatorTracker(
+            ThreadInfoRequestCoordinator coordinator,
+            GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
+            Function<OperatorThreadInfoStats, T> createStatsFn,
+            ExecutorService executor,
+            int cleanUpInterval,
+            int numSamples,
+            int statsRefreshInterval,
+            Time delayBetweenSamples,
+            int maxStackTraceDepth) {
+
+        this.coordinator = checkNotNull(coordinator, "Thread info samples coordinator");
+        this.resourceManagerGatewayRetriever =
+                checkNotNull(resourceManagerGatewayRetriever, "Gateway retriever");
+        this.createStatsFn = checkNotNull(createStatsFn, "Create stats function");
+        this.executor = checkNotNull(executor, "Scheduled executor");
+
+        checkArgument(cleanUpInterval >= 0, "Clean up interval");
+
+        checkArgument(numSamples >= 1, "Number of samples");
+        this.numSamples = numSamples;
+
+        checkArgument(
+                statsRefreshInterval >= 0,
+                "Stats refresh interval must be greater than or equal to 0");
+        this.statsRefreshInterval = statsRefreshInterval;
+
+        this.delayBetweenSamples = checkNotNull(delayBetweenSamples, "Delay between samples");
+
+        checkArgument(
+                maxStackTraceDepth >= 0,
+                "Max stack trace depth must be greater than or equal to 0");
+        this.maxThreadInfoDepth = maxStackTraceDepth;
+
+        this.operatorStatsCache =
+                CacheBuilder.newBuilder()
+                        .concurrencyLevel(1)
+                        .expireAfterAccess(cleanUpInterval, TimeUnit.MILLISECONDS)
+                        .build();
+    }
+
+    @Override
+    public Optional<T> getOperatorStats(AccessExecutionJobVertex vertex) {
+        synchronized (lock) {
+            final T stats = operatorStatsCache.getIfPresent(vertex);
+            if (stats == null
```

Comment:
Where do we prevent multiple requests being initiated when the first one has not completed yet? Asking since the check is based on the operatorStatsCache that is only populated once the response from the TaskExecutor is received.

Reply:
https://github.com/apache/flink/pull/15054/files/e7f59e2e2811c0718a453572e90a4ad2a900ff03#diff-c9e5dc5e2c21890cb66f2592f45b3b23462fd28f330a86a1aec824576fb60039R301  
(`ThreadInfoOperatorTracker` line 301)